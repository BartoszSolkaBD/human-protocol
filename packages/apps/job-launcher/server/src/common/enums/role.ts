export enum Role {
  Exchange = 'exchange',
  Recording = 'recording',
  Reputation = 'reputation',
  JobLauncher = 'job-launcher',
}
