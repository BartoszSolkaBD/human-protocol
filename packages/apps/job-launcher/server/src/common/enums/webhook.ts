export enum EventType {
  ESCROW_CREATED = 'escrow_created',
  ESCROW_CANCELED = 'escrow_canceled',
  SUBMISSION_REJECTED = 'submission_rejected',
}

export enum OracleType {
  FORTUNE = 'fortune',
  CVAT = 'cvat',
  HCAPTCHA = 'hcaptcha',
}

export enum WebhookStatus {
  PENDING = 'PENDING',
  COMPLETED = 'COMPLETED',
  FAILED = 'FAILED',
}
