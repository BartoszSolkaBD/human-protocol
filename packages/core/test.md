1. Can we use viaIR with foundry?

Contracts are compuling with viaIR using Foundry. 
You specify via-ir=true in Foundry.toml

2. If we start using viaIR, can we still upgrade deployed proxies?

Contracts were upgraded using viaIR in fact using Foundry. 


3. If we use viaIR, how will it affect to the gas cost?

Using viaIR, gas cost decreased slightly. 
Use : forge snapshot --diff with and without via-ir=true in Foundry.toml


4. If we move to Foundry, can we still upgrade deployed proxies?

Yes, deployed proxies were upgraded using Foundry. 