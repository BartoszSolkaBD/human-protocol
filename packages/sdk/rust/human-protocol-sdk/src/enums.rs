use chrono::{DateTime, Utc};
use thiserror::Error;
use serde_derive::{Deserialize, Serialize};

/// Represents various blockchain network IDs.
#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq, Hash)]
pub enum ChainId {
    Mainnet = 1,
    Rinkeby = 4,
    Goerli = 5,
    BscMainnet = 56,
    BscTestnet = 97,
    Polygon = 137,
    PolygonMumbai = 80001,
    Moonbeam = 1284,
    MoonbaseAlpha = 1287,
    Avalanche = 43114,
    AvalancheTestnet = 43113,
    Skale = 1273227453,
    Localhost = 1338
}

/// Represents the status of an escrow.
#[derive(Debug, Clone, PartialEq)]
pub enum EscrowStatus {
    /// Escrow is launched.
    Launched,
    /// Escrow is funded, and waiting for the results to be submitted.
    Pending,
    /// Escrow is partially paid out.
    Partial,
    /// Escrow is fully paid.
    Paid,
    /// Escrow is finished.
    Complete,
    /// Escrow is cancelled.
    Cancelled,
}

impl EscrowStatus {
    pub fn to_string_vec() -> Vec<String> {
        vec![
            String::from("Launched"),
            String::from("Pending"),
            String::from("Partial"),
            String::from("Paid"),
            String::from("Complete"),
            String::from("Cancelled"),
        ]
    }
}

pub enum Role {
  JobLauncher,
  ExchangeOracle,
  RecordingOracle,
  ReputationOracle
}

impl Role {
    pub fn to_string_vec() -> Vec<String> {
        vec![
            String::from("Job Launcher"),
            String::from("Exchange Oracle"),
            String::from("Recording Oracle"),
            String::from("Reputation Oracle")
        ]
    }
}


#[derive(Debug, Error, PartialEq, Eq)]
pub enum EthereumError {
    #[error("Invalid Ethereum Address Error")]
    InvalidEthereumAddressError,
    #[error("Transaction Failed Error")]
    TransactionFailedError,
    #[error("Invalid Argument Error")]
    InvalidArgumentError,
    #[error("Web3 Error: {0}")]
    Web3Error(String),
    #[error("Web3 Contract Error: {0}")]
    Web3ContractError(String),
    #[error("Web3 Event Error: {0}")]
    Web3EventError(String),
    #[error("Event Parsing Error")]
    EventParsingError,
    #[error("Invalid Fee Error")]
    InvalidFeeError,
    #[error("Invalid Url Error")]
    InvalidUrlError,
    #[error("Invalid Key Error")]
    InvalidKeyError,
    #[error("Empty Hash Error")]
    EmptyHashError,
    #[error("Invalid Hash Error {0} != {1}")]
    InvalidHashError(String, String),
    #[error("Escrow Address Not Provided By Factory Error")]
    EscrowAddressNotProvidedByFactoryError,
    #[error("Unknown Escrow Status Error")]
    UnknownEscrowStatusError,
    #[error("Escrow Has Invalid Status Error")]
    EscrowHasInvalidStatusError,
    #[error("Invalid Recipient Size Error")]
    InvalidRecipientsSizeError,
    #[error("Invalid Amounts Size Error")]
    InvalidAmountsSizeError,
    #[error("Invalid Trusted Handlers Size Error")]
    InvalidTrustedHandlersSizeError,
    #[error("Recipients And Amounts Must Be Same Length Error")]
    RecipientsAndAmountsMustBeSameLengthError,
    #[error("Invalid Keys Size Error")]
    InvalidKeysSizeError,
    #[error("Invalid Values Size Error")]
    InvalidValuesSizeError,
    #[error("Keys And Values Must Be Same Length Error")]
    KeysAndValuesMustBeSameLengthError,
    #[error("Escrow Does Not Have Enough Balance Error")]
    EscrowDoesNotHaveEnoughBalanceError,
    #[error("Invalid ChainId Error")]
    InvalidChainIdError,
    #[error("Invalid Dates Error: {0} must be earlier than {1}")]
    InvalidDatesError(DateTime<Utc>, DateTime<Utc>),
    #[error("Network Not Found For Chain Id Error: {0}")]
    NetworkNotFoundForChainIdError(u32),
    #[error("Escrow Not Found For Escrow Address Error: {0}")]
    EscrowNotFoundForEscrowAddressError(String),
    #[error("No Logs Found Error")]
    NoLogsFoundError,
    #[error("Invalid Token Address Found Error")]
    InvalidTokenAddressFoundError,
    #[error("Log Parsing Error: {0}")]
    LogParsingError(String),
    #[error("Escrow Cancel Error")]
    EscrowCancelError,
    #[error("Unexpected Status Code Error {0}")]
    UnexpectedStatusCodeError(u16),
    #[error("Request Error {0}")]
    RequestError(String),
    #[error("Non Positive Value Error")]
    NonPositiveValueError,
    #[error("Rewards Not Found For Slasher Address Error: {0}")]
    RewardsNotFoundForSlasherAddressError(String),
    #[error("Leaders Not Found Error")]
    LeadersNotFoundError,
    #[error("Leader Not Found Error: {0}")]
    LeaderNotFoundError(String),
}

#[derive(Debug, Error, PartialEq, Eq)]
pub enum Error {
    #[error("Invalid Token Conversion Error")]
    InvalidTokenConversionError,
    #[error("Invalid Tuple Length Error")]
    InvalidTupleLengthError
    
}
