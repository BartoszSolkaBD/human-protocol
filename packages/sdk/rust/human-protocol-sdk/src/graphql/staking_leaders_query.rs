use cynic::{impl_scalar};
use serde::{Serialize};


use crate::enums::{Role};

mod schema {
  cynic::use_schema!("schemas/leader.graphql");
}

impl_scalar!(i64, schema::I64);

#[derive(Serialize, Debug, Clone, cynic::QueryFragment)]
#[cynic(schema_path = "schemas/leader.graphql")]
pub struct Leader {
  id: String,
  address: String,
  amount_staked: cynic::Id,
  amount_allocated: cynic::Id,
  amount_locked: cynic::Id,
  locked_until_timestamp: cynic::Id,
  amount_withdrawn: cynic::Id,
  amount_slashed: cynic::Id,
  reputation: cynic::Id,
  reward: cynic::Id,
  amount_jobs_launched: cynic::Id,
  role: Option<String>,
  fee: Option<cynic::Id>,
  public_key: Option<String>,
  webhook_url: Option<String>,
  url: Option<String>,
}

#[derive(cynic::QueryVariables, Clone)]
pub struct LeaderFilter {
  role: Option<Vec<String>>
}


impl Default for LeaderFilter {
  fn default() -> Self {
    Self {
      role: Some(Role::to_string_vec()),
    }
  }
}

impl LeaderFilter {
  pub fn role(mut self, role: Option<Vec<String>>) -> Self {
    self.role = role;
    self
  }

  pub fn build(self) -> LeaderFilter {
    LeaderFilter {
      role: self.role,
    }
  }
}

#[derive(cynic::QueryFragment, Debug)]
#[cynic(graphql_type = "Root", variables = "LeaderFilter")]
#[cynic(schema_path = "schemas/leader.graphql")]
pub struct GetLeadersQuery {
  #[arguments(
    where: { 
      role_in: $role
    }
  )]
  pub leaders: Option<Vec<Leader>>,
}

pub async fn run_query(url: &str, filter: &LeaderFilter) -> cynic::GraphQlResponse<GetLeadersQuery> {
  use cynic::http::ReqwestExt;

  let query = build_query(filter);
  
  reqwest::Client::new()
    .post(url)
    .run_graphql(query)
    .await
    .unwrap()
}

fn build_query(filter: &LeaderFilter) -> cynic::Operation<GetLeadersQuery, LeaderFilter> {
  use cynic::QueryBuilder;

  GetLeadersQuery::build(filter.clone())
}

#[cfg(test)]
mod test {
    use crate::{enums::ChainId, constants::NETWORKS};

    use super::*;
    use tokio::test;

    #[test]
    async fn snapshot_test_query() {
      let filter = LeaderFilter::default();
      let query = build_query(&filter);

      insta::assert_snapshot!(query.query);
    }

    #[tokio::test]
    async fn test_running_query() {
      let network = NETWORKS.get(&ChainId::PolygonMumbai).unwrap();
      let filter = LeaderFilter::default();
      let result = run_query(network.subgraph_url, &filter).await;
      if let Some(errors) = &result.errors {
          assert_eq!(errors.len(), 0);
      }
      insta::assert_debug_snapshot!(result.data);
    }

    #[tokio::test]
    async fn test_filter_builder() {
        let filter = LeaderFilter::default()
            .role(Some(vec!["some_role".to_string()]))
            .build();

        assert_eq!(filter.role, Some(vec!["some_role".to_string()]));
    }
}