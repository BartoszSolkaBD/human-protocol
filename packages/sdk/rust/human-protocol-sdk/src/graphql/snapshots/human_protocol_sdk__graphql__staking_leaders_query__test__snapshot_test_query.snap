---
source: src/graphql/staking_leaders_query.rs
expression: query.query
---
query GetLeadersQuery($role: [String!]) {
  leaders(where: {role_in: $role, }) {
    id
    address
    amountStaked
    amountAllocated
    amountLocked
    lockedUntilTimestamp
    amountWithdrawn
    amountSlashed
    reputation
    reward
    amountJobsLaunched
    role
    fee
    publicKey
    webhookUrl
    url
  }
}


